### ISSUE TITLE
Figure out index_status / index_wait interface

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#3663]
We have changed the interface for what used to be `table_config`, `table_status`, `table_wait` to `table.config()`, `table.status()` and `table.wait()` in 1.16 (https://github.com/rethinkdb/rethinkdb/issues/3365).

However we didn't have time to agree on a consistent interface for indexes, which as of now still use `index_status`, `index_wait` etc.
We should make this consistent. 

Some ideas from the other thread:
@mlucy 
> * Introduce `table.index()`.
  - If nothing is chained after it, `table.index('a')` returns the opaque representation of `a`.
    * This lets us duplicate indexes by writing `r.table('test').index_create('a_dup', r.table('test').index('a'))` rather than having to pull the information out of `index_status`.
  - Add `table.index('a').status()`, `table.index('a').wait` by analogy to the `table.X` commands.
[...]
If people don't like that, another possibility would be `r.table('test').status(index:'a')` rather than `index_status`.

@larkost 
> I like the idea of "embedding" the index data in with the table since I feel that indexes are conceptually a part of their tables. I was a little surprised that `.table_wait()` did not wait for indexes to be ready (makes sense if you think about indexes as part of the table, but not if you see them as separate). So changing this would agree with my intuitive understanding of it.
>    
> Adding an `indexes` opt-arg to the `.wait()` command to specify what you are waiting for would be a solution, e.g.:
>    
>  ```python
r.db('test').table('a').wait().run() # wait for everything, including all indexes on the table
r.db('test').table('a').wait(indexes=False).run() # wait for the table to be read/write available, but not the indexes
r.db('test').table('a').wait(indexes=['b']).run() # wait for the table to be read/write available, and the index 'b' to be available
```
>
>  I don't think it makes sense to wait for indexes but not some form of read or write availability, and I am not sure about a multi-table form that makes sense in this case.

**Comment by timmaxw** [comment]
I'd forbid `r.table(...).index(...)` with nothing chained after it, because of the analogy with tables. For tables, `r.table(...)` returns contents and `r.table(...).config()` returns metadata. The index's opaque representation is metadata, not content. So I'd make `r.table(...).index(...)` produce an error and ``r.table(...).index(...).config()` return a document with the opaque representation. We could also make `.config()` return a writable `name` field to replace `index_rename()`.

For `.wait()`, I'd forbid `r.table(...).wait(indexes=[...])` and instead have the user do `r.table(...).index(...).wait()` if they want to wait on a specific index. I think it's rare that people want to wait on more than one index but not all of the indexes, so I don't see a reason to offer a special syntax for it.

I think this can wait for v2.1.

**Comment by deontologician** [comment]
Would it be bad if `table().index(..)` returned the actual values in the index? I've seen people ask how they can get that data before. It might be an intuitive value for the index content.

**Comment by timmaxw** [comment]
I don't think we should do that for 2.0, but maybe in the long term.

**Comment by gchpaco** [comment]
I like @mlucy's proposed ReQL interface, it has a pleasing symmetry with the table stuff and the ability to e.g. dupe indexes easily might be helpful.

The `wait` issue it seems is a general limitation; what we want is some ability to say, idiomatically, "I want to wait until all these things are done."  Union waiting is easy—just issue them in sequence—but we may want to make it easier on the user to specify them; maybe `wait` on a sequence type so I could do something like `[r.table(...), r.table(...), r.table(...).index(...), ...].wait()`.

**Comment by danielmewes** [comment]
I'm 100% on board with what @timmaxw wrote.

I think if we have the `indexes` optarg to `table.wait(indexes = <boolean>)` in addition to being able to call `table.index(...).wait()` we would cover all important cases.

I'm not generally opposed to supporting wait() on a sequence, though I think it's not very important. Even without it `for_each` can be used I think to implement waiting on multiple things, and it's not that complicated to write in (the probably rare) case where a user needs it.

It actually seems like this might become a significant amount of work. Let's keep discussing the proposals, but I might consider moving it over to 2.1 later.

**Comment by mlucy** [comment]
> `[r.table(...), r.table(...), r.table(...).index(...), ...].wait()`

I agree; that seems like a useful thing to want to do.  That particular syntax wouldn't work right now because you can't store tables in arrays, but `r.wait(table1, table2, index, ...)` would work, and would probably be a good idea.

---

I think supporting `r.table('test').index('a').wait()` but erroring on `r.table('test').index('a')` would be really bad.  People find ReQL's chaining intuitive because they can build commands up piece by piece, and because they can have a clear mental model of what every sub-chain returns.  Having queries that produce errors by themselves, but which produce something useful when you chain something else on the end is weird.  (`default` is a special case because it's meant to handle errors.)

I would much prefer for `r.table('test').index('a')` to return an opaque representation of the index.  It makes the query language feel more consistent, it's strictly more useful, and I don't think anyone's going to be confused by it.  (Or if they are confused, I don't think they'll be any *more* confused than if they got an error in the same situation.)

**Comment by timmaxw** [comment]
I don't think we should have any special support for waiting on a sequence. We already support waiting on a specific index; all of a table's indexes; the table itself; or all the tables in a database. In the unlikely event that the user wants to do something more complicated, like wait on two indexes without waiting on a third one, they can do something like `[...].foreach(f(r.row).wait())`, or just call `wait()` on the things one at a time. So we already make the common things easy and the uncommon things possible. @mlucy @gchpaco Why do you think waiting on a sequence is common enough to justify its own special syntax?

@mlucy, I think you have a good point that it's nice to be able to build up to `r.table('test').index('a').wait()` piece by piece. But I think it would be confusing for `r.table('test').index('a')` to return a binary blob with no explanation. Can we think of something for `.index()` to return that will make users say "that's an index" when they see it? Maybe we could return the same thing as `.index().config()`, but maybe not writable.

**Comment by mlucy** [comment]
@timmaxw -- does waiting on a table wait on its indexes?

We could add an index pseudotype in place of a binary blob for the return status of `r.table('test').index('a')`.  We could also return the value of `index.status`.  If we returned the value of `index.config`, I think it would be weird for it not to be writable, but I'd be cool with making `.index` a shortcut for `.index.config`.  (And, of course, we could always just return a wrapper object like `{function: BLOB}` instead of `BLOB`.)

Do any of those four sound palatable?

**Comment by deontologician** [comment]
We do have `r.db('foo')` as well if we're insisting on everything having a meaning.

**Comment by timmaxw** [comment]
Currently, waiting on a table does not wait on its indexes. But I'm fine with the proposal to support `table.wait(indexes=true|false)`. Probably the default should be `true`.

I don't think an index pseudo-type is the right way to go. Instead I'd add a new value type (like TABLE) that supports `.wait()`, `.config()`, etc. but that's coercible to a datum, where the datum coercion is equivalent to `.status()`, or perhaps a string of the form `<index 'foo' on 'bar.baz'>`.

But I think @deontologician makes a good point that `r.db("foo")` without anything chained onto it returns a (fairly impenetrable!) error message, and this doesn't seem to have confused anybody.

**Comment by mlucy** [comment]
`r.db('foo')` is a bit of a special case.  It does return a legal value, it's just not a value we return to users (mostly because we didn't have pseudotypes when we were first implementing ReQL).  So you get back an error, but it's a "Query result cannot be of type DATABASE" error -- it wasn't evaluating `r.db(...)` that gave you the error, it was trying to return it to yourself.

That said, I'd be up for a proposal to make `r.db(...)`'s return value less confusing.

**Comment by mlucy** [comment]
I could be convinced to make `index` return a value like DATABASE that we never send back to the user.  That would still let us support `index_create('a', r.table('test').index('b'))` for example, and feels much less weird to me than saying that `.index('a')` is an error and `.index('a').create()` isn't.

**Comment by mlucy** [comment]
(To be clear, I would still prefer to pick a reasonable representation that we *can* return to the user -- I think that's the least confusing option.)

**Comment by timmaxw** [comment]
I'd prefer to make it like DATABASE but improve the error message for returning a DATABASE/INDEX/whatever to the user. One reason is that we might eventually make INDEX do something useful, like return the index values, and if we don't make it coercible to datum then we can do that later without breaking compatibility.

**Comment by timmaxw** [comment]
Also, I think it would be slightly confusing to have two separate ways to get an index's status, through `.index(...)` and through `.index(...).status(...)`.

**Comment by timmaxw** [comment]
Another problem with returning a representation to the user is that implicit conversions are confusing.

**Comment by mlucy** [comment]
OK, that seems like a reasonable compromise, and we might well want to make `index` convertible to `stream` in the future (the way `table` is already).

**Comment by mlucy** [comment]
Marking this as settled with the interpretation that `.index()` will return a type `INDEX` which can be used in index creation but not returned to the user (much like `DB`).  We'll chain `status` and `config` after `index`.

**Comment by danielmewes** [comment]
The API looks fine. But I think we should do this in 2.1 not 2.0 since it's a bit more than a trivial amount of work.

**Comment by larkost** [comment]
Pinging this one, as `table.wait()` not covering indexes came up again.
